File: src/github.py
from requests.models import Response
from api_client import ApiClient
from type.github import IGithubLatestReleases


class GitHub:
    repo_url: str
    user_name: str
    user_repo: str
    github_endpoint: str
    user_agent: str
    api_client: ApiClient

    def __init__(self, repo_url: str) -> None:
        self.repo_url = repo_url
        self.user_name = self.repo_url.split('/')[3]
        self.user_repo = self.repo_url.split('/')[4]
        self.github_endpoint = 'https://api.github.com'
        self.user_agent = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.36'

        self.api_client = ApiClient()

    def get_the_latest_release(self) -> IGithubLatestReleases:
        self.repo_url = f'{self.github_endpoint}/repos/{self.user_name}/{self.user_repo}/releases/latest'

        headers = {
            'User-Agent': self.user_agent
        }

        return self.api_client.get(self.repo_url, headers=headers).json()

    def get_tags(self) -> Response:
        self.repo_url = f'{self.github_endpoint}/repos/{self.user_name}/{self.user_repo}/tags'

        headers = {
            'User-Agent': self.user_agent
        }

        return self.api_client.get(self.repo_url, headers=headers).json()

File: modules/github.go
package modules

import (
	"chobb/helpers"
	"chobb/types"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"path"
	"strings"
)

var UserAgent string = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.36"

func GetTheLatestRelease(RepoUrl string) types.GitHubLatestReleases {
	var SplittedRepo []string = strings.Split(RepoUrl, "/")
	var UserName string = SplittedRepo[3]
	var UserRepo string = SplittedRepo[4]
	RepoUrl = fmt.Sprintf("https://api.github.com/repos/%s/%s/releases/latest", UserName, UserRepo)

	RequestClient := &http.Client{}
	req, _ := http.NewRequest("GET", RepoUrl, nil)
	req.Header.Set("User-Agent", UserAgent)

	resp, err := RequestClient.Do(req)

	if err != nil {
		log.Fatalln("Could not fetch gh latest release", err)
	}

	file, _ := ioutil.ReadAll(resp.Body)
	defer resp.Body.Close()
	var GithubLatestRelease types.GitHubLatestReleases
	json.Unmarshal(file, &GithubLatestRelease)

	return GithubLatestRelease
}

func Download(DownloadUrl string, FileName string) {
	RequestClient := &http.Client{}
	req, _ := http.NewRequest("GET", DownloadUrl, nil)
	req.Header.Set("User-Agent", UserAgent)

	resp, err := RequestClient.Do(req)

	if err != nil {
		log.Fatalln("Could not download", FileName)
		helpers.BgDangerMessage("Could not download")
		return
	}

	var File string = path.Join(helpers.ChobPath(), FileName)
	out, err := os.Create(File)
	defer out.Close()

	io.Copy(out, resp.Body)
	err = os.Chmod(File, 0777)

	if err != nil {
		log.Fatalln("Could not set 777 chmod!")
		return
	}
	// exec.Command("chmod +x %s", File).Start()
	defer resp.Body.Close()
}
